core/src/test/java/com/radixdlt/api/core/TransactionAccuTreeTest.java

```diff
@@ -0,0 +1,134 @@
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.api.core;
+
+import static com.radixdlt.harness.predicates.NodesPredicate.nodeAt;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.hash.HashCode;
+import com.radixdlt.api.DeterministicCoreApiTestBase;
+import com.radixdlt.crypto.HashUtils;
+import com.radixdlt.harness.predicates.NodePredicate;
+import com.radixdlt.lang.Option;
+import com.radixdlt.transaction.REv2TransactionAndProofStore;
+import java.util.stream.LongStream;
+import org.junit.Test;
+
+public class TransactionAccuTreeTest extends DeterministicCoreApiTestBase {
+
+  @Test
+  public void stateManagerMaintainsCorrectTransactionMerkleTree() {
+    try (var test = buildRunningServerTest(6)) {
+      // Run the setup until 2 epoch proofs are captured
+      var reader = test.getInstance(0, REv2TransactionAndProofStore.class);
+      test.runUntilState(nodeAt(0, NodePredicate.atOrOverEpoch(1)), 1000);
+      var genesisHeader = reader.getEpochProof(1).get().ledgerHeader();
+      test.runUntilState(nodeAt(0, NodePredicate.atOrOverEpoch(2)), 1000);
+      var epochHeader = reader.getEpochProof(2).get().ledgerHeader();
+
+      // Capture the transactions between these 2 proofs
+      var epochTransactions =
+          LongStream.range(
+                  genesisHeader.accumulatorState().stateVersion().toNonNegativeLong().unwrap() + 1,
+                  epochHeader.accumulatorState().stateVersion().toNonNegativeLong().unwrap() + 1)
+              .mapToObj(reader::getTransactionAtStateVersion)
+              .map(Option::unwrap)
+              .toList();
+
+      // Assert a certain count (on which we rely during latter manual merkle computation)
+      assertThat(epochTransactions.size()).isEqualTo(6);
+
+      // Capture the transaction hashes
+      var transactionHashes =
+          epochTransactions.stream()
+              .map(transaction -> HashUtils.blake2b256(transaction.transactionBytes()))
+              .toArray(HashCode[]::new);
+
+      // Assert that header's root hash is equal to manually computed one
+      assertThat(epochHeader.hashes().transactionRoot())
```

Comment:
Out of interest, what half-empty tree abstraction do we use? Are all empty/childless nodes given HashUtils.zero256()?

Reply:
The placeholder hash "cuts" the tree branch short (not going all the way to the leaves).
I have now updated the test data to demonstrate that 👍